text/0000-declarative-root.md

```diff
@@ -0,0 +1,280 @@
+<template>
+    <!-- Comment -->
+    <lwc:root></lwc:root>
+</template>
+```
+
+However, in the case of `lwc:preserve-comments`, comments may not precede `<lwc:root>`:
+
+```html
+<!-- Invalid -->
+<template lwc:preserve-comments>
+    <!-- Comment -->
+    <lwc:root></lwc:root>
+</template>
+```
+
+#### Contents
+
+The `<lwc:root>` element may not have any contents other than whitespace and comments:
+
+```html
+<!-- Valid -->
+<template>
+    <lwc:root>
+        
+    </lwc:root>
+</template>
+```
+
+```html
+<!-- Valid -->
+<template>
+    <lwc:root>
+        <!-- Comment -->
+    </lwc:root>
+</template>
+```
+
+However, in the case of `lwc:preserve-comments`, comments inside of `<lwc:root>` are not allowed:
+
+```html
+<!-- Invalid -->
+<template lwc:preserve-comments>
+    <lwc:root>
+        <!-- Comment -->
+    </lwc:root>
+</template>
+```
+
+This restriction is why `lwc:inner-html` is not supported on `<lwc:root>`.
+
+### Timing
+
+In terms of timing, any attributes added to or removed from the root element should follow the timing applied
+to elements inside of the template:
+
+```html
+<template>
+    <lwc:root class={clazz}></lwc:root>
+    <div class={clazz}></div>
+</template>
+```
+
+In other words, in the above example, if `clazz` changes, then the root element's `class` attribute should be changed
+in the same tick as when the `<div>`'s `class` changes.
+
+The above also applies to the timing of when event listeners are attached using `on*`.
+
+No guarantees are made about the ordering of changes made to `<lwc:root>` compared to the ordering of changes made to elements inside of the template. I.e. there is no guarantee that changes to the root occur before changes to an in-template element, or vice-versa.
+
+The ordering of changes made to the element itself (i.e. between the categories of `class`, other attributes, and event listeners) is also not guaranteed. (Nor is it guaranteed for in-template elements in any other RFC.)
+
+### Precedence
+
+Attributes and event listeners added using `<lwc:root>` may conflict with those added by the parent component:
+
+```html
+<!-- parent.html -->
+<template>
+    <x-child class="foo" 
+             data-foo="bar" 
+             onclick={onClick} 
+             lwc:spread={others}>
+    </x-child>
+</template>
+```
+
+```html
+<!-- child.html -->
+<template>
+    <lwc:root class="quux" 
+              data-foo="toto" 
+              onclick={onClick} 
+              lwc:spread={others}>
+    </lwc:root>
+</template>
+```
+
+In cases of conflict, the following rules apply:
+
+- `lwc:spread` precedence applies [as normal](https://github.com/salesforce/lwc-rfcs/pull/52) within the `<lwc:root>` element itself.
+- For event listeners such as `onclick`, both event listeners are attached.
+- For the `class` attribute, strings are concatenated with a single whitespace (`" "`) character (e.g. `"foo quux"` in the above example). No attempt is made to deduplicate duplicate strings.

```

Comment:
How do you envision reconciliating the `class` attribute if it is changed programmatically?

For example, what would be the generated `class` attribute value if the parent runs this in its `renderedCallback` and if the child updates the class on the `<lwc:root class={computed}>`?

```
this.template.querySelector('x-child').classList.add('prog')
```

IMO, it's reasonable to discard any programmatic change if the `class` property is set declaratively.

Reply:
We have this same problem today already with scoped styles. Classes are added by the framework to every element.

I agree with you that we should not make any attempt to reconcile programmatic changes. My philosophy is "you broke it, you bought it." Developers should be discouraged from programmatically modifying things that the framework is also modifying.

FWIW, Svelte takes this same approach with their scoped styles – there is no attempt to merge programmatic styles with framework-added styles:

![Screen Shot 2023-01-20 at 11 12 04 AM](https://user-images.githubusercontent.com/283842/213785883-96298383-6fef-4cd4-9f16-8609a2a1c502.png)

Reply:
2d4109e